crates/types/src/horizon.rs

```diff
@@ -1,96 +1,47 @@
 pub struct Horizon {
-    pub left_horizon_y: f32,
-    pub right_horizon_y: f32,
+    pub point_on_horizon: Point2<Pixel>,
+    pub horizon_normal: Vector2<Pixel>,
 }
 
 impl Horizon {
     pub fn horizon_y_minimum(&self) -> f32 {
-        self.left_horizon_y.min(self.right_horizon_y)
+        self.y_at_x(0.0).min(self.y_at_x(640.0))
     }
 
-    pub fn y_at_x(&self, x: f32, image_width: f32) -> f32 {
-        self.left_horizon_y + x / image_width * (self.right_horizon_y - self.left_horizon_y)
+    pub fn y_at_x(&self, x: f32) -> f32 {
+        let normal = self.horizon_normal;
+        let point = self.point_on_horizon;
+
+        -normal.x() * (x - point.x()) / normal.y() + point.y()
     }
 
     pub fn from_parameters(
-        camera_to_ground: Transform<Camera, Ground, Isometry3<f32>>,
-        focal_length: Vector2<f32>,
-        optical_center: Point2<f32>,
-        image_width: f32,
+        camera_to_ground: Isometry3<Camera, Ground>,
+        intrinsics: Matrix3x4<f32>,
     ) -> Self {
-        let rotation_matrix = camera_to_ground.inner.rotation.to_rotation_matrix();
-        let horizon_slope_is_infinite = rotation_matrix[(2, 2)] == 0.0;
-
-        if horizon_slope_is_infinite {
-            Self {
-                left_horizon_y: 0.0,
-                right_horizon_y: 0.0,
-            }
-        } else {
-            let left_horizon_y = optical_center.y
-                + focal_length.y
-                    * (rotation_matrix[(2, 0)]
-                        + optical_center.x * rotation_matrix[(2, 1)] / focal_length.x)
-                    / rotation_matrix[(2, 2)];
-            let slope = -focal_length.y * rotation_matrix[(2, 1)]
-                / (focal_length.x * rotation_matrix[(2, 2)]);
-
-            // Guesses if image size is in "normalized" (1.0 x 1.0) dimensions
-            let adjusted_image_width = if image_width <= 1.0 {
-                image_width
-            } else {
-                image_width - 1.0
-            };
-            let right_horizon_y = left_horizon_y + (slope * adjusted_image_width);
-
-            Self {
-                left_horizon_y,
-                right_horizon_y,
-            }
+        let horizon_normal = Vector3::z_axis();
+        let horizon_normal_camera = camera_to_ground.inverse() * horizon_normal;
+        let horizon_normal_camera: Vector3<Pixel> =
+            vector![horizon_normal_camera.x(), horizon_normal_camera.y(), 0.0].normalize();
+        let horizon_normal_image = intrinsics * horizon_normal_camera.inner.to_homogeneous();
+
+        let camera_front = Vector3::z_axis();
+        let ground_front = camera_to_ground * camera_front;
+        let ground_front = vector![ground_front.x(), ground_front.y(), 0.0].normalize();
+
+        let horizon_point_camera = camera_to_ground.inverse() * ground_front;
+        let horizon_point_image = intrinsics * horizon_point_camera.inner.to_homogeneous();
```

Comment:
These variables are called `point`, but are vectors.

Reply:
Weeeell, hard problem, these are points that are infinitely far away in ground. Renamed to vanishing_point to make it clearer